doc/SAI-Proposal-Debug-Counters.md

```diff
@@ -0,0 +1,264 @@
+# Generic SAI Debug Counters
+
+Title       | Generic SAI Debug counters
+------------|----------------
+Authors     | Mellanox
+Status      | In review
+Type        | Standards track
+Created     | 07/17/2019
+SAI-Version | 1.5
+
+## Overview
+ASICs can provide a set of debug counters for certain object types.
+Debug counters belong to certain families, each family is for cetain object type.
+The content of a specific debug counter instance can be defined by the application.
+Counting every statistics of every family might be too resource intensive, therefore the debug counters provide an efficient way to count and aggregate only the needed information.
+
+## Debug counter object type
+For the counter object, a sai_counter_stat_t enum is defined:
+```
+typedef enum _sai_debug_counter_type_t
+{
+    /** Port in drop reasons. Base object : SAI_OBJECT_TYPE_PORT */
+    SAI_DEBUG_COUNTER_TYPE_PORT_IN_DROP_REASONS,
+
+    /** Port out drop reasons. Base object : SAI_OBJECT_TYPE_PORT */
+    SAI_DEBUG_COUNTER_TYPE_PORT_OUT_DROP_REASONS,
+
+} sai_debug_counter_type_t;
+```
+
+## Debug counter port in drop reason family
+```
+typedef enum _sai_port_in_drop_reason_t
+{
+    /* L2 reasons */
```

Comment:
The current enums would need changes to add drop-reasons as newer ASICs are added. This would keep constantly changing. An alternative is to do this generically and let vendors add new drop reasons without having to change SAI spec or SONIC.

Say, we have two switch read-only attributes  SAI_SWITCH_ATTR_INGRESS_PORT_DROP_REASON_LIST and 
SAI_SWITCH_ATTR_EGRESS_PORT_DROP_REASON_LIST. NOS queries these attributes and platform lists the drop port stat ids it supports along with descriptive names. NOS would just query the port stats with these ids and print the port, descriptive name of the drop reason, and the drop counter value.

typedef struct _sai_drop_reason_t
{
    /** drop reason Id */
    sai_uint32_t id;

    /** drop reason name */
    char reason_name[128];

} sai_drop_reason_t;

typedef struct _sai_drop_reason_list_t
{
    /** Number of entries in the list */
    uint32_t count;

    /**drop reason list */
    sai_drop_reason_t *list;

} sai_drop_reason_list_t;


//     * @type sai_switch_drop_reason_list_t
//      * @flags READ_ONLY
SAI_SWITCH_ATTR_INGRESS_PORT_DROP_REASON_LIST
SAI_SWITCH_ATTR_EGRESS_PORT_DROP_REASON_LIST

Reply:
IMO, you can have the same comment on any SAI feature, such as all counters, or even all attributes. Sure, there are many differences between different ASICs. So we try to capture a common basis as the standard, and allow any vendor to extend the list with custom range (I will go ahead and add that to the proposal)
Having a completely custom list takes all the approach behind unified API, and makes SAI just a dispatcher for vendor specific API

Reply:
IMO, debug_counters are different from other counters. Here, the goal is to not have silent packet drops. The SONIC code we are discussing here today really doesn't need to actually understand the drop_reason to do anything. The SONIC code just needs  to
1) let the user know about the debug counters supported by the platform
2) let the user configure different combination of drop counters on a given port identify the actual drop reason.
3) drop reasons vary widely across ASICs and there are no standards for it as against port stats etc.

Reply:
in the spec, we show example to query the drop reasons for a platform. what you can do is to add your own list of reason to the enum and once a nos query it will use those enum supported by your platform for debugging.